hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/append/AppendWriteFunctionWithBufferSort.java

```diff
@@ -0,0 +1,111 @@
+
+import org.apache.flink.table.data.binary.BinaryRowData;
+import org.apache.flink.table.runtime.operators.sort.BinaryInMemorySortBuffer;
+import org.apache.flink.table.runtime.util.MemorySegmentPool;
+import org.apache.flink.util.MutableObjectIterator;
+import org.apache.hudi.configuration.FlinkOptions;
+import org.apache.hudi.sink.StreamWriteOperatorCoordinator;
+import org.apache.hudi.sink.append.sort.RowDataComparator;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.hudi.sink.buffer.MemorySegmentPoolFactory;
+import org.apache.hudi.sink.utils.BufferUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Sink function to write the data to the underneath filesystem with buffer sort
+ * to improve the parquet compression rate.
+ *
+ * <p>The function writes base files directly for each checkpoint,
+ * the file may roll over when it’s size hits the configured threshold.
+ *
+ * @param <T> Type of the input record
+ * @see StreamWriteOperatorCoordinator
+ */
+public class AppendWriteFunctionWithBufferSort<T> extends AppendWriteFunction<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(AppendWriteFunctionWithBufferSort.class);
+  private final long writebufferSize;
+  private final RowDataComparator rowDataComparator;
+  private transient BinaryInMemorySortBuffer buffer;
+  private transient MemorySegmentPool memorySegmentPool;
+
+  public AppendWriteFunctionWithBufferSort(Configuration config, RowType rowType) {
+    super(config, rowType);
+    this.writebufferSize = config.get(FlinkOptions.WRITE_BUFFER_SIZE);
+    String sortKeys = config.get(FlinkOptions.WRITE_BUFFER_SORT_KEYS);
+    if (sortKeys == null) {
+      throw new IllegalArgumentException("Sort keys can't be null for append write with buffer sort.");
+    }
+
+    this.rowDataComparator = new RowDataComparator(rowType, sortKeys);
+  }
+
+  @Override
+  public void open(Configuration parameters) throws Exception {
+    super.open(parameters);
+    this.memorySegmentPool = MemorySegmentPoolFactory.createMemorySegmentPool(config);
+    this.buffer = BufferUtils.createBuffer(rowType, memorySegmentPool, rowDataComparator);
+  }
+
+  @Override
+  public void processElement(T value, Context ctx, Collector<Object> out) throws Exception {
+    RowData data = (RowData) value;
+    buffer.write(data);
+    if (buffer.size() >= writebufferSize) {
+      sortAndSend();
+    }
+  }
+
+  @Override
+  public void snapshotState() {
+    try {
+      sortAndSend();
+    } catch (IOException e) {
+      LOG.error("Fail to sort and flush data in buffer during snapshot state.");
+      throw new FlinkRuntimeException(e);
+    }
+    super.snapshotState();
+  }
+
+  private void sortAndSend() throws IOException {
```

Comment:
Seems the parquet file may contains records from multiple `sortAndSend`, so the written file is partially ordered on sort keys?

Reply:
Generally, sorting implies that all data within a file is strictly ordered. This means each sortAndSend operation typically creates a new file. Of course, we can control the buffer memory size to minimize the proliferation of small files as much as possible.

Reply:
Yes, wholly ordered holds more significance. For append writing, the flushing can be triggered with two conditions:
* Checkpoint trigger. in which current remaining data in buffer are flushed and committed.
* Binary buffer is full. We can set the size of buffer as the max size of underly file, e.g., PARQUET_MAX_FILE_SIZE, to keep aligned with current behavior considering the file's size / quantity.

Reply:
Agree. Added onto method comment.